File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] edge) {
        int answer = 0;
        
        // 시작점 1로 고정 bfs
        
        List<List<Integer>> graph = new ArrayList<>();
        for(int i=0; i<=n; i++) graph.add(new ArrayList<>());
        
        for(int[] e : edge){
            graph.get(e[0]).add(e[1]);
            graph.get(e[1]).add(e[0]);
        }
        
        Queue<Integer> q = new LinkedList<>();
        q.add(1);
        
        int[] dist = new int[n+1];
        Arrays.fill(dist, -1);
        dist[1] = 0;
        
        int maxnum = 0;
        
        while(!q.isEmpty()){
            int cur = q.poll();
            
            for(int g : graph.get(cur)){
                if(dist[g] != -1) continue;
                dist[g] = dist[cur]+1;
                q.offer(g);
                maxnum = Math.max(dist[g], maxnum);
            }
        }
        for(int i=1; i<=n; i++){
            if(dist[i] == maxnum) answer++;
        }
        
        return answer;
    }
}

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.cpp
#include <bits/stdc++.h>
using namespace std;

// bfs 
vector<int> v[50005];
int dist[50005];
bool check[50005] = {0, };

int solution(int n, vector<vector<int>> edge) {
     // 양방향 인접리스트
    for(int i=0; i<edge.size(); i++){
        int a = edge[i][0];
        int b = edge[i][1];
        v[a].push_back(b);
        v[b].push_back(a);
    }
    
    
    queue<int> q;
    q.push(1);
    check[1]=1;
    dist[1] =0;
    int i =0;
    while(!q.empty()){
        int cur = q.front();
        q.pop();
        cout<<'\n';
        for(auto c : v[cur]){
            if(check[c]) continue;
            dist[c] = (dist[cur]+1);
            check[c]=1;
            q.push(c);
            i= max(i, dist[c]);
        }
    }
    int total=0;
    for(int j=1; j<=n; j++){
        if(dist[j] == i) total++;
    }
    return total;
}